drivers/can/can_common.c

```diff
@@ -35,3 +46,100 @@ int z_impl_can_attach_msgq(struct device *dev, struct k_msgq *msg_q,
 
 	return api->attach_isr(dev, can_msgq_put, msg_q, filter);
 }
+
+static inline void can_work_buffer_init(struct can_frame_buffer *buffer)
+{
+	buffer->head = 0;
+	buffer->tail = 0;
+}
+
+static inline int can_work_buffer_put(struct zcan_frame *frame,
+				      struct can_frame_buffer *buffer)
+{
+	u16_t next_head = WORK_BUF_MOD_SIZE(buffer->head + 1);
+
+	if (buffer->head == WORK_BUF_FULL) {
+		return -1;
+	}
+
+	buffer->buf[buffer->head] = *frame;
+
+	/* Buffer is almost full */
+	if (next_head == buffer->tail) {
+		buffer->head = WORK_BUF_FULL;
+	} else {
+		buffer->head = next_head;
+	}
+
+	return 0;
+}
+
+static inline
+struct zcan_frame *can_work_buffer_get_next(struct can_frame_buffer *buffer)
+{
+	/* Buffer empty */
+	if (buffer->head == buffer->tail) {
+		return NULL;
+	} else {
+		return &buffer->buf[buffer->tail];
+	}
+}
+
+static inline void can_work_buffer_free_next(struct can_frame_buffer *buffer)
+{
+	u16_t next_tail = WORK_BUF_MOD_SIZE(buffer->tail + 1);
+
+	if (buffer->head == buffer->tail) {
+		return;
+	}
+
+	if (buffer->head == WORK_BUF_FULL) {
+		buffer->head = buffer->tail;
+	}
+
+	buffer->tail = next_tail;
+}
+
+static void can_work_handler(struct k_work *work)
+{
+	struct zcan_work *can_work = CONTAINER_OF(work, struct zcan_work,
+						  work_item);
+	struct zcan_frame *frame;
+
+	while ((frame = can_work_buffer_get_next(&can_work->buf))) {
+		can_work->cb(frame, can_work->cb_arg);
+		can_work_buffer_free_next(&can_work->buf);
+	}
+}
+
+static void can_work_isr_put(struct zcan_frame *frame, void *arg)
+{
+	struct zcan_work *work = (struct zcan_work *)arg;
+	int ret;
+
+	ret = can_work_buffer_put(frame, &work->buf);
+	if (ret) {
+		LOG_ERR("Workq buffer overflow. Msg ID: 0x%x",
+			frame->id_type == CAN_STANDARD_IDENTIFIER ?
+				frame->std_id : frame->ext_id);
+		return;
+	}
+
+	k_work_submit_to_queue(work->work_queue, &work->work_item);
+}
+
+int z_impl_can_attach_workq(struct device *dev, struct k_work_q *work_q,
+			    struct zcan_work *work,
+			    can_rx_callback_t callback, void *callback_arg,
+			    const struct zcan_filter *filter)
+{
+	const struct can_driver_api *api = dev->driver_api;
+
+	k_work_init(&work->work_item, can_work_handler);
+	work->work_queue = work_q;
+	work->cb = callback;
+	work->cb_arg = callback_arg;
+	can_work_buffer_init(&work->buf);
+
+	return api->attach_isr(dev, can_work_isr_put, work, filter);
```

Comment:
We have a security problem here.
The pointer to the zcan_work came from user memory. The problem is we will be using this data structure when the ISR fires. But what is going to stop the user thread from concurrently overwriting the members of the zcan_work struct with garbage or malicious data? If that happens, terrible things might happen can_work_isr_put() actually runs depending on what was changed, particularly the work->work_queue pointer.

Reply:
@andrewboie I could check the data within the ISR. The user can't run concurrently to an ISR.

Reply:
An alternative would be making the zcan_work a kernel object. But how can I copy the zcan_frame to the userspace then?

Reply:
@andrewboie Userspace handler for attaching workqueues is removed until workqueues get proper userspace support.

Reply:
since there is no handler can you remove the __syscall

Reply:
Removed.